File: relatime-spark/src/main/scala/com/star/gmall/realtime/util/PhoenixUtil.scala
package com.star.gmall.realtime.util

import java.sql.{DriverManager, PreparedStatement, ResultSet, ResultSetMetaData}

object PhoenixUtil {
  val url = "jdbc:phoenix:node:2181"
  /**
  * 执行 sql, 并封装查询后的结果到Map中
  */
  def query(sql: String, args: List[Object]): List[Map[String, Object]] = {
    var result = List[Map[String, Object]]( )

    val conn = DriverManager.getConnection(url)
    println("sql:"+sql)
    val ps: PreparedStatement = conn.prepareStatement(sql)
    (1 to args.size).foreach(i => {
      ps.setObject(i, args(i - 1))
    })
    val resultSet: ResultSet = ps.executeQuery()
    val meta: ResultSetMetaData = resultSet.getMetaData // 需要知道有多少列
    while (resultSet.next()) {
      var map: Map[String, Object] = Map[String, Object]()
      for (i <- 1 to meta.getColumnCount) { // 遍历每一列
        val name = meta.getColumnName(i).toLowerCase()
        val value = resultSet.getObject(i)
        map += name -> value
      }
      result :+= map
    }
    ps.close()
    conn.close()
    result
  }

}

File: gmall-realtime/src/main/java/com/star/gmall/realtime/utils/PhoenixUtil.java
package com.star.gmall.realtime.utils;

import com.star.gmall.realtime.common.GmallConfig;
import net.minidev.json.JSONObject;
import org.apache.commons.beanutils.BeanUtils;
import org.codehaus.jackson.map.util.BeanUtil;

import java.lang.reflect.InvocationTargetException;
import java.sql.*;
import java.util.ArrayList;
import java.util.List;

public class PhoenixUtil {
    public static Connection conn = null;

    public static void queryInit() {
        try {
            Class.forName("org.apache.phoenix.jdbc.PhoenixDriver");
            conn = DriverManager.getConnection(GmallConfig.PHOENIX_SERVER);
            conn.setSchema(GmallConfig.HBASE_SCHEMA);
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
        } catch (SQLException throwables) {
            throwables.printStackTrace();
        }
    }

    public static <T> List<T> queryList(String sql,Class<T> clazz) {
        if (conn == null) {
            queryInit();
        }

        ArrayList<T> resultList = new ArrayList<>();
        PreparedStatement ps = null;
        ResultSet rs = null;
        try {
            ps = conn.prepareStatement(sql);
            rs = ps.executeQuery();
            ResultSetMetaData md = rs.getMetaData();
            while (rs.next()) {
                T rowData = clazz.newInstance();
                for (int i = 1; i < md.getColumnCount(); i++) {
                    BeanUtils.setProperty(rowData,md.getColumnName(i),rs.getObject(i));
                }
                resultList.add(rowData);
            }
        } catch (SQLException throwables) {
            throwables.printStackTrace();
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        } catch (InstantiationException e) {
            e.printStackTrace();
        } catch (InvocationTargetException e) {
            e.printStackTrace();
        }
        return resultList;
    }


    public static void main(String[] args) {
        List<JSONObject> jsonObjects = queryList("select * from base_trademark", JSONObject.class);
        System.out.println(jsonObjects);
    }
}
